File: jawa/yoda/src/main/java/ru/prohor/universe/yoda/bot/handlers/command/HelpCommandHandler.java
package ru.prohor.universe.yoda.bot.handlers.command;

import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.objects.Message;
import ru.prohor.universe.jocasta.tgbots.api.FeedbackExecutor;
import ru.prohor.universe.jocasta.tgbots.api.comand.CommandHandler;
import ru.prohor.universe.yoda.bot.Text;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class HelpCommandHandler implements CommandHandler {
    private final List<CommandHandler> commandHandlers;

    public HelpCommandHandler(List<CommandHandler> commandHandlers) {
        List<CommandHandler> mutable = new ArrayList<>(commandHandlers);
        mutable.add(this);
        this.commandHandlers = mutable;
    }

    @Override
    public String command() {
        return "/help";
    }

    @Override
    public String description() {
        return Text.CommandDescriptions.help();
    }

    @Override
    public boolean handle(Message message, FeedbackExecutor feedbackExecutor) {
        feedbackExecutor.sendMessage(
                message.getChatId(),
                commandHandlers.stream()
                        .map(cmd -> cmd.command() + " - " + cmd.description())
                        .collect(Collectors.joining("\n")) // TODO
        );
        return false;
    }
}

File: jawa/yoda/yoda-bot/src/main/kotlin/handlers/command/HelpCommandHandler.kt
package ru.prohor.universe.yoda.bot.handlers.command

import org.springframework.stereotype.Service
import org.telegram.telegrambots.meta.api.objects.Message
import ru.prohor.universe.jocasta.core.collections.common.Opt
import ru.prohor.universe.jocasta.tgbots.api.FeedbackExecutor
import ru.prohor.universe.jocasta.tgbots.api.comand.CommandHandler
import ru.prohor.universe.yoda.bot.Text.Commands
import ru.prohor.universe.yoda.bot.Text.CommandDescriptions

@Service
class HelpCommandHandler(commandHandlers: List<CommandHandler>) : CommandHandler {
    private val commandsListMessage by lazy {
        createMessage(commandHandlers)
    }

    override fun command(): String {
        return Commands.HELP
    }

    override fun description(): Opt<String> {
        return Opt.of(CommandDescriptions.help())
    }

    override fun handle(message: Message, feedbackExecutor: FeedbackExecutor) {
        feedbackExecutor.sendMessage(
            message.chatId,
            commandsListMessage
        )
    }

    private fun createMessage(handlers: List<CommandHandler>): String {
        return handlers
            .plus(this)
            .filter { cmd -> cmd.description().isPresent }
            .joinToString(separator = "\n") { cmd ->
                "${cmd.command()} - ${cmd.description().get()}"
            }
    }
}
